src/persist/src/storage.rs

```diff
@@ -141,14 +141,61 @@ pub enum Atomicity {
     AllowNonAtomic,
 }
 
-/// An abstraction over a `bytes key`->`bytes value` store.
+/// An abstraction over read-only access to a `bytes key`->`bytes value` store.
+#[async_trait]
+pub trait BlobRead: Send + 'static {
+    /// Returns a reference to the value corresponding to the key.
+    async fn get(&self, key: &str) -> Result<Option<Vec<u8>>, Error>;
+
+    /// List all of the keys in the map.
+    async fn list_keys(&self) -> Result<Vec<String>, Error>;
+
+    /// Synchronously closes the blob, causing all future commands to error.
+    ///
+    /// If `Self` also implements Read, this releases the exclusive-writer lock.
+    ///
+    /// Implementations must be idempotent. Returns true if the blob had not
+    /// previously been closed.
+    ///
+    /// NB: It's confusing for this to be on BlobRead, since it's a no-op on the
+    /// various concrete {Mem,File,S3}Read impls, but in various places we need
+    /// to be able to close something that we only know is a BlobRead. Possible
+    /// there's something better we could be doing here.
+    async fn close(&mut self) -> Result<bool, Error>;
+}
+
+/// An abstraction over read-write access to a `bytes key`->`bytes value` store.
+///
+/// Blob and BlobRead impls are allowed to be concurrently opened for the same
+/// location in the same process (which is often used in tests), but this is not
+/// idiomatic for production usage. Instead, within a process, only single Blob
+/// or BlobRead impl should be used for each unique storage location.
```

Comment:
I don't know if we need to specify this in this doc, but when we say unique storage location we mean "unique and not nested" or something

e.g. 

/foo and /foo/bar do not count as two unique storage locations i think. right?

Reply:
this just sounds like an invalid usage?